lib/src/rules/component_annotation_arguments_ordering.dart

```diff
@@ -0,0 +1,233 @@
+
+    return componentAnnotation != null
+        ? _visitAnnotation(componentAnnotation)
+        : [];
+  }
+
+  List<_ArgumentInfo> _visitAnnotation(Annotation node) {
+    final _argumentsInfo = <_ArgumentInfo>[];
+
+    final arguments = node.arguments.arguments.whereType<NamedExpression>();
+
+    for (final argument in arguments) {
+      final name = argument.name.label.name;
+      final group = _ArgumentGroup.parseArgumentName(name);
+
+      if (group != null && _groupsOrder.contains(group)) {
+        _argumentsInfo.add(_ArgumentInfo(
+          argument: argument,
+          argumentOrder: _getOrder(group, name, _argumentsInfo),
+        ));
+      }
+    }
+
+    return _argumentsInfo;
+  }
+
+  _ArgumentOrder _getOrder(
+    _ArgumentGroup group,
+    String argumentName,
+    Iterable<_ArgumentInfo> argumentsInfo,
+  ) {
+    if (argumentsInfo.isNotEmpty) {
+      final lastArgumentOrder = argumentsInfo.last.argumentOrder;
+      final hasSameGroup = lastArgumentOrder.argumentGroup == group;
+
+      final previousArgumentGroup = hasSameGroup
+          ? lastArgumentOrder.previousArgumentGroup
+          : lastArgumentOrder.argumentGroup;
+
+      return _ArgumentOrder(
+        argumentGroup: group,
+        previousArgumentGroup: previousArgumentGroup,
+        isWrong: (hasSameGroup && lastArgumentOrder.isWrong) ||
+            _isCurrentGroupBefore(lastArgumentOrder.argumentGroup, group),
+      );
+    }
+
+    return _ArgumentOrder(isWrong: false, argumentGroup: group);
+  }
+
+  bool _isCurrentGroupBefore(
+    _ArgumentGroup lastArgumentGroup,
+    _ArgumentGroup argumentGroup,
+  ) =>
+      _groupsOrder.indexOf(lastArgumentGroup) >
+      _groupsOrder.indexOf(argumentGroup);
+
+  bool _isComponentAnnotation(Annotation node) =>
+      node.name.name == 'Component' && node.atSign.type.lexeme == '@';
+}
+
+@immutable
+class _ArgumentGroup {
+  final String name;
+  final Iterable<String> arguments;
+
+  static const selector = _ArgumentGroup._(
+    'selector',
+    ['selector'],
+  );
+  static const templates = _ArgumentGroup._(
+    'templates',
+    ['template', 'templateUrl'],
+  );
+  static const styles = _ArgumentGroup._(
+    'styles',
+    ['styles', 'styleUrls'],
+  );
+  static const directives = _ArgumentGroup._(
+    'directives',
+    ['directives', 'directiveTypes'],
+  );
+  static const pipes = _ArgumentGroup._(
+    'pipes',
+    ['pipes'],
+  );
+  static const providers = _ArgumentGroup._(
+    'providers',
+    ['providers', 'viewProviders'],
+  );
+  static const encapsulation = _ArgumentGroup._(
+    'encapsulation',
+    ['encapsulation'],
+  );
+  static const visibility = _ArgumentGroup._(
+    'visibility',
+    ['visibility'],
+  );
+  static const exports = _ArgumentGroup._(
+    'exports',
+    ['exports', 'exportAs'],
+  );
+  static const changeDetection = _ArgumentGroup._(
+    'change_detection',
+    ['changeDetection'],
+  );
+
+  static const _groupsOrder = [
```

Comment:
@incendial I think it's a not good idea have two _groupsOrder. (first int class ComponentAnnotationArgumentsOrderingRule and in global scope)

Reply:
@dkrutskikh why not? First one in `ComponentAnnotationArgumentsOrderingRule` is private, and can't be accessed outside the class. Second one is available only through `_ArgumentGroup` class (which is private). Could u please explain more?